File: Week1/strings/longest_substring_without_repeat.py
def longestSubstringWithoutDuplication(string):
    #hold position of recent appearance of each letter
    lastSeen = {}

    #longest substring with startindex and end index. 
    longest = [0, 1]

    startIdx = 0

    for i, char in enumerate(string):
        if char in lastSeen:
            startIdx = max(startIdx, lastSeen[char] + 1)
        
        if longest[1] - longest[0] < i + 1 - startIdx:
            longest = [startIdx, i + 1]

        lastSeen[char] = i
    
    return string[longest[0]:longest[1]]


print(longestSubstringWithoutDuplication('Bardia'))

File: Week1/strings/longestSubstringWithoutRepeat.js
//takes in a string as input
//outputs the longest substring without duplication

//e.g.

//input = 'clementisacap'
//output = 'mentisac'


//solution


const longestSubstringWithoutDuplication = (string) => {
    const lastSeen = {}
    //longest as in an array that contains the starting and ending indeces of the longest substring
    let longest = [0, 1];
    let startIndex = 0;


    for (let i = 0; i < string.length; i++) {
        let currentChar = string[i];

        if (lastSeen[currentChar] !== undefined) {
            startIndex = Math.max(startIndex, lastSeen[currentChar] + 1)
        }

        if (longest[1] - longest[0]  < i + 1  - startIndex) {
            longest = [startIndex, i + 1]
        }

        lastSeen[currentChar] = i
    }

    return string.substring(longest[0], longest[1])
}



console.log(longestSubstringWithoutDuplication("clementisac"))

